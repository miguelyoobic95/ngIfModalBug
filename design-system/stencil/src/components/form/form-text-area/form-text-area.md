---
name: FormTextArea
category: Core Components
---

```form-text-area.html
    <yoo-form-text-area></yoo-form-text-area>
```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`text`|string|   |the text inside the label|

## CSS

|Type|Name|Description|
|---|---|---|
|Accent|`accent`|form-text-area accent|
|Dark|`dark`|form-text-area dark|
|Success|`success`|form-text-area success|
|Danger|`danger`|form-text-area danger|
|Info|`info`|form-text-area info|
|Warning|`warning`|form-text-area warning|
|Gradient Accent|`gradient-accent`|form-text-area gradient-accent|
|Gradient Dark|`gradient-dark`|form-text-area gradient-grey|
|Gradient Success|`gradient-success`|form-text-area gradient-success|
|Gradient Danger|`gradient-danger`|form-text-area gradient-danger|
|Gradient Info|`gradient-info`|form-text-area gradient-info|
|Gradient Warning|`gradient-warning`|form-text-area gradient-warning|

## Events
|Attr|Description|
|---|---|---|
|`yooXXXX`|small label|