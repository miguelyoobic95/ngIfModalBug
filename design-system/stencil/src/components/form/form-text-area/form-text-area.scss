:host {
    textarea {
        width: 100%;
        height: 100%;
        min-height: 95px;
        border: solid 0.0625rem cssvar(stable-30);
        border-radius: 0.3125rem;
        color: cssvar(dark);
        font-size: 1.0625rem;
        font-weight: normal;
        padding: 0.5rem 1rem;
        &.invalid {
            border-color: cssvar(danger);
        }
        &.valid {
            border-color: cssvar(success);
        }
        &:focus {
            outline: none;
        }
    }
}