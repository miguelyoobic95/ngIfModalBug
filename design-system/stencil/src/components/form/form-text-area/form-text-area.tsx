import { Component, Element, Prop, Event, EventEmitter, Method, State } from '@stencil/core';
import { ValidatorEntry, AsyncValidator, Validator, IFormInputBase } from '@shared/interfaces';
import { setValidator, setAsyncValidator, onInputBlurred, setValueAndValidateInput, onInputFocused } from '../../../utils/helpers/form-input-helpers';

@Component({
    tag: 'yoo-form-text-area',
    styleUrl: 'form-text-area.scss',
    scoped: true
})
export class YooFormTextAreaComponent implements IFormInputBase<string | number> {

    @Prop({ mutable: true }) value: string;
    @Prop() validators: Array<Validator<string> | ValidatorEntry> = [];
    @Prop() asyncValidators: Array<AsyncValidator<string>>;
    @Prop() placeholder: string;
    @Prop() readonly: boolean;
    @Prop() resizable: 'both' | 'vertical' | 'horizontal' | 'none' = 'both';

    @Event() validityChanged: EventEmitter<boolean>;
    @Event() inputBlurred: EventEmitter<any>;
    @Event() inputFocused: EventEmitter<boolean>;
    @Event() inputChanged: EventEmitter<any>;

    @State() validity: boolean;

    @Element() host: HTMLStencilElement;

    // Reduced Validators
    _validator: Validator<string> = (x: string) => true;
    _asyncValidator: AsyncValidator<string> = async (x: string) => true;

    componentWillLoad() {
        setValidator(this.validators);
        setAsyncValidator(this.asyncValidators);
    }

    componentDidLoad() {
        if (this.resizable !== 'both') {
            let textArea = this.host.querySelector('textarea');
            textArea.setAttribute('style', `resize: ${this.resizable};`);
        }
    }

    onInputChanged(ev: any): void {
        let value = ev.target.value;
        setValueAndValidateInput(value, this);
    }

    @Method()
    isValid() {
        return this.validity;
    }

    renderEditable(): JSX.Element {
        return (
            <textarea value={this.value}
                placeholder={this.placeholder}
                readonly={this.readonly}
                onBlur={ev => onInputBlurred(ev, this, 'textarea')}
                onInput={ev => this.onInputChanged(ev)}
                onFocus={ev => onInputFocused(ev, this, 'textarea')}
            />
        );
    }

    renderReadonly() {
        return <div innerHTML={this.value}></div>;
    }

    render(): JSX.Element {
        return this.readonly ? this.renderReadonly() : this.renderEditable();
    }
}
