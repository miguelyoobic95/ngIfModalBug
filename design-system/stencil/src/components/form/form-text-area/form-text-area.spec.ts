import { TestWindow } from '@stencil/core/testing';
import { YooFormTextAreaComponent } from './form-text-area';

import { generateEvent, inputFieldValue } from '../../../../test';


describe('FormTextAreaComponent', () => {
    let window;
    beforeEach(() => {
      window = new TestWindow();
    });

    it('should build', () => {
        expect(new YooFormTextAreaComponent()).toBeTruthy();
    });

    describe('Snapshots', () => {
        let element;

        it ('Should render', async () => {
            element = await window.load({
                components: [YooFormTextAreaComponent],
                html: '<yoo-form-text-area></yoo-form-text-area>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('Should render with a placeholder', async () => {
            element = await window.load({
                components: [YooFormTextAreaComponent],
                html: '<yoo-form-text-area placeholder="placeholder"></yoo-form-text-area>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('Should render with text', async () => {
            element = await window.load({
                components: [YooFormTextAreaComponent],
                html: '<yoo-form-text-area value="text"></yoo-form-text-area>'
            });
            expect(element).toMatchSnapshot();
        });

    });

    describe('Events', () => {

        let element, res, input;

        beforeEach( async () => {
            element = await window.load({
                components: [YooFormTextAreaComponent],
                html: '<yoo-form-text-area></yoo-form-text-area>'
            });
            input = element.querySelector('textarea');
        });

        it('should emit inputChanged', async() => {
            let inputSpy = jest.fn(ev => {res = ev.detail; });
           window.document.addEventListener('inputChanged', inputSpy);
           const value = 'Text area input value';
           inputFieldValue(input, value);
           expect(inputSpy).toHaveBeenCalled();
           expect(res).toEqual(value);
        });

        it(' should emit blur and focus event', async () => {
            let blurSpy = jest.fn(ev => {res = ev.detail; });
            let focusSpy = jest.fn(ev => {res = ev.detail; });
            window.document.addEventListener('inputBlurred', blurSpy);
            window.document.addEventListener('inputFocused', focusSpy);
            await input.dispatchEvent(generateEvent(window, 'focus'));
            window.flush();
            expect(focusSpy).toHaveBeenCalled();
            await input.dispatchEvent(generateEvent(window, 'blur'));
            window.flush();
            expect(blurSpy).toHaveBeenCalled();
        });
    });
});