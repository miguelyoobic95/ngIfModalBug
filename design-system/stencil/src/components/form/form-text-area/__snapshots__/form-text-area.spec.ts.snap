// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextAreaComponent Snapshots Should render 1`] = `
<yoo-form-text-area
  class="hydrated"
  data-yoo-form-text-area-host=""
>
  
  <textarea
    data-yoo-form-text-area=""
  />
</yoo-form-text-area>
`;

exports[`FormTextAreaComponent Snapshots Should render with a placeholder 1`] = `
<yoo-form-text-area
  class="hydrated"
  data-yoo-form-text-area-host=""
  placeholder="placeholder"
>
  
  <textarea
    data-yoo-form-text-area=""
    placeholder="placeholder"
  />
</yoo-form-text-area>
`;

exports[`FormTextAreaComponent Snapshots Should render with text 1`] = `
<yoo-form-text-area
  class="hydrated"
  data-yoo-form-text-area-host=""
  value="text"
>
  
  <textarea
    data-yoo-form-text-area=""
  />
</yoo-form-text-area>
`;
